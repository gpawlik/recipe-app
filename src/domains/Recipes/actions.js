import { actionTypes as at } from './constants';

export const fetch = query => ({
  type: at.RECIPES_FETCH,
  query
});

export const fetchSuccess = data => ({
  type: at.RECIPES_FETCH_SUCCESS,
  data
});

export const fetchError = error => ({
  type: at.RECIPES_FETCH_ERROR,
  error
});
