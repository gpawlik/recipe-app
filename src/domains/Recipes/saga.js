import { call, put, fork, takeLatest } from 'redux-saga/effects';

import { fetchRecipes } from './service';
import { fetchSuccess, fetchError } from './actions';
import { actionTypes as at} from './constants';

export function* fetchAllRecipes({ query }) {
  try {
    const result = yield call(fetchRecipes, query);

    yield put(fetchSuccess(result));
  } catch (err) {
    yield put(fetchError(err));
  }
}

export function* recipesWatcher() {
  yield fork(takeLatest, at.RECIPES_FETCH, fetchAllRecipes);
}
