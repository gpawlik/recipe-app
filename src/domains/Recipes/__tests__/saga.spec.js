import { put } from 'redux-saga/effects';

import { fetchSuccess, fetchError } from '../actions';
import { fetchAllRecipes } from '../saga';

const query = 'foo';

describe('fetchRecipes Saga', () => {
  let fetchRecipesGenerator;

  beforeEach(() => {
    fetchRecipesGenerator = fetchAllRecipes({ query });

    const callDescriptor = fetchRecipesGenerator.next(query).value;

    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the fetchSuccess action if it requests the data successfully', () => {
    const response = [{
      title: 'foo'
    }];
    const putDescriptor = fetchRecipesGenerator.next(response).value;

    expect(putDescriptor).toEqual(put(fetchSuccess(response, query)));
  });

  it('should call the fetchError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = fetchRecipesGenerator.throw(response).value;

    expect(putDescriptor).toEqual(put(fetchError(response)));
  });
});
