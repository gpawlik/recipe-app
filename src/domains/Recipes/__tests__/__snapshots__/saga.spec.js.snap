// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchRecipes Saga should call the fetchError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "foo",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`fetchRecipes Saga should dispatch the fetchSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "foo",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
